### ISSUE TITLE
Support for showing tags for collapsed spans

### ISSUE THREAD

**Issue by bobrik** [jaegertracing/jaeger-ui#675]
## Requirement - what kind of business use case are you trying to solve?

I'm trying to reduce the number of distinct operations and make it easier to aggregate and analyze traces with many spans of the same type, while retaining the ability to see what happens in the trace without expanding too many spans.

## Problem - what in Jaeger blocks you from solving the requirement?

I have a trace with many similar operations like "load module by name" or "run command X with args", here's an example:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/89186/104133873-25f63380-533b-11eb-9a66-c50958a3c99b.png">

This leads to many distinct operations, since every module gets its own. Trace stats are suffering from this, not allowing me to do group by (operation, module name), because both things are lumped into one string of `operationName`:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/89186/104134031-0c092080-533c-11eb-8aea-b7bbfc27dc30.png">

It also makes it hard to search for "any module load above Xms". It's still doable if I attach the same tag to all module loading operations, but it doesn't feel right.

If I were to not include the module name into operation name, the trace would've looked like this:

![image](https://user-images.githubusercontent.com/89186/104133928-753c6400-533b-11eb-8f41-30c9d961f85e.png)

This is much harder to understand, since I have 2000+ spans for module loading and I'm trying to figure out whether some of them are getting loaded for no reason.

## Proposal - what do you suggest to solve the problem or improve the existing situation?

I suggest to designate some tag to be showed in the timeline:

![image](https://user-images.githubusercontent.com/89186/104134085-686c4000-533c-11eb-940d-ecefa3dd804b.png)

I think it solves all the outlined problems at expense of requiring a new tag to be attached, which feels like an okay compromise. For spans that do not opt into having a new tag it's a no-op.

## Any open questions to address

I'm happy to consider other options.

**Comment by yurishkuro** [comment]
I agree this is a valuable feature. Another common situation where it can be useful is on outbound HTTP calls, which are often represented simply with span name `GET` because instrumentation does not have an easy way of extracting low-cardinality name from the URL. But if we had a feature like this, those spans could be annotated with the full URL just in the display.

The thing we need to design is how the UI should know which tag to pull into display for different types of spans (NB: somewhat relevant discussion in OpenTelemetry: https://github.com/open-telemetry/opentelemetry-specification/pull/810). Some options are:

1. A dictionary that maps (service name, span name) -> display pattern, where pattern is a string that uses placeholders for tag names, e.g. `"${tag_name}"`. The downside of this approach is it can get tedious specifying this for many services, and it does not allow solving for HTTP spans generically
2. A variant of (1) where instead of exact service/span name key there is a matcher that describes the spans generically, e.g. `span.kind = client AND span.tags.http.url != null`. This will require more careful design of the matcher DSL (maybe we could use something off the shelf, such as https://jsonlogic.com/)
3. An option that was discussed (and eventually rejected) in OpenTelemetry is to introduce a magic tag like `display.name` to a span. I don't like it for the reasons similar to objections raised in OTEL.

I feel that option (2) is the way to go, it's the most flexible & expressive, and can be fully controlled from the UI configuration (instead of forcing instrumentation to worry about display issues as in option (3)).

**Comment by bobrik** [comment]
Option 2 sounds reasonable to me.

Since it's something that is controlled by the owner of the UI, should we consider allowing JS instead of a DSL? It can then not only extract the name, but also annotate the span, e.g.:

* `load_module` span can link to the source code for the module
* `get_user` span can link to an admin page for the user
* `rpc` span can link to a dashboard specific for this rpc
* a warning can be attached to a span it it breaches SLO, and JS can look up SLO dynamically

There are probably other use cases that are application specific. It may be something that jaeger-ui explicitly doesn't want to incorporate and that's ok, but it's worth thinking about the limits of customization.

**Comment by yurishkuro** [comment]
Support for JS config is coming (https://github.com/jaegertracing/jaeger/pull/2707).

The linking from span tags to external pages is already supported: https://www.jaegertracing.io/docs/1.21/frontend-ui/#link-patterns

**Comment by bobrik** [comment]
I'll see if I can prototype something when jaegertracing/jaeger#2707 is merged, it seems like a good fit.

**Comment by jkowall** [comment]
@bobrik any update on this?

**Comment by bobrik** [comment]
Some feedback in #678 would be nice.
